### PR TITLE
refactor(auth): Make TokenService a singleton via lru_cache

### LINKED ISSUE #164: Code quality: Minor cleanup items from assessment
## Summary

Minor code quality issue identified during the 2026-01-11 assessment. Low priority but worth tracking.

## Item

| Issue | Location | Notes |
|-------|----------|-------|
| Eager DB initialization | `auth/service.py:66` | TokenService initializes DB even if unused |

## Resolution

**Original concern:** TokenService calls `init_database()` eagerly in `__init__`, even if the service is never used.

**Actual fix:** Making `get_token_service()` return a singleton via `@functools.lru_cache(maxsize=1)` prevents redundant `TokenService` construction. Since the service is only constructed once per application lifetime, the eager `init_database()` call in `__init__` is fine.

**What didn't work:** Lazy initialization with double-checked locking was over-engineering. The problem was multiple TokenService instances per request, not eager init within each instance. The singleton pattern solves the real issue.

## Priority

Low - this is a minor improvement, not a bug.

## Related

- Broad exception catch split to #238
- Missing type hint addressed in #217

---
(AI-generated via Claude Code w/ Opus 4.6)

### PR DESCRIPTION
**Pull request by duplico**
## Summary

Makes `get_token_service()` return a cached singleton via `@functools.lru_cache`, preventing redundant `TokenService` construction and `init_database()` calls on every request.

**What changed:**
- Added `@functools.lru_cache(maxsize=1)` to `get_token_service()` in `deps.py`
- Added `clear_token_service_cache()` for test teardown
- Singleton test coverage

**What was removed (from earlier iterations):**
- Double-checked locking pattern in `service.py` — solved a problem that doesn't exist
- Class-level `_initialized_paths` tracking — unnecessary with singleton
- Per-method `_ensure_database_initialized()` calls — eager init in `__init__` is fine for a singleton
- Thread-safety machinery — `@lru_cache` handles this

Closes #164

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by duplico** [comment]
Round 1 complete: PR is merge-ready. No review comments found, all CI checks passing.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/magpie/auth/service.py [hunk 1]
```diff
@@ -94,8 +94,17 @@ def __init__(self, config: MagpieSettings) -> None:
         """
         self.config = config
         self.db_path = config.database_path
-        # Ensure database is initialized
-        init_database(self.db_path)
+        self._db_initialized = False
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/auth/service.py]: Fixed in f97d4d0. Added @functools.lru_cache(maxsize=1) to get_token_service() to match the pattern used by get_settings(). This caches the TokenService instance as a singleton, preventing repeated database initialization on every request.

(AI-generated via Claude Code w/ Claude Sonnet 4.5)

**Comment by duplico** [comment]
## Round 1 Complete

**Comments addressed**: 1/1
**Commits**: f97d4d0

### Actions Taken
- Fixed inefficient database initialization pattern in `get_token_service()`
- Added `@functools.lru_cache(maxsize=1)` decorator to cache TokenService as singleton
- Prevents database initialization on every request (was idempotent but inefficient)
- Solution matches existing pattern used by `get_settings()`

### Results
- CI: All checks passing (lint, test, security, e2e)
- Tests: All 847 unit tests passed

### Remaining
- 0 unresolved threads
- Next: Adversarial review before merge

(AI-generated via Claude Code w/ Sonnet 4.5)

**Comment by duplico** [comment]
## Adversarial Review: PASS

No significant correctness concerns found. PR is ready for merge.

### Verification Chain
- Issue #164 requirement: Lazy DB initialization to avoid init when TokenService is unused
- PR implementation: Correctly implements lazy init via `_ensure_database_initialized()`
- Code verification: All 4 DB-accessing methods properly call lazy init
- Tests: Verify lazy behavior + all 847 unit tests passing

### Integration Impact
- Backward compatible (API surface unchanged)
- `magpie-ctl init` still works (explicit init before TokenService usage)
- FastAPI singleton via `@lru_cache` ensures single instance across requests

### Minor Observations (Not Blocking)
- Thread-safety of `_db_initialized` flag: Not lock-protected, but `init_database()` is idempotent (CREATE TABLE IF NOT EXISTS), so concurrent calls are safe
- This aligns with PR goal: avoid init when unused, not optimize concurrent init

**Status**: merge_ready

(AI-generated via Claude Code w/ Opus 4)

File: src/magpie/auth/service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/auth/service.py]: Fixed in f97d4d0. Added `@functools.lru_cache(maxsize=1)` to `get_token_service()` in `src/magpie/server/deps.py` to ensure the TokenService instance is cached and reused across requests, preventing repeated database initialization.

The fix ensures:
- Database initialization happens only once per application lifecycle
- All requests share the same TokenService instance
- No performance impact from repeated DB init

Verified that all unit tests pass.

(AI-generated via Claude Code w/ Sonnet 4.5)

**Comment by duplico** [comment]
Starting final validation. Found 1 unresolved review comment - addressing now.

**Status**: CI passing, 1 review comment to address

**Comment to address**:
- tests/unit/test_token_service.py:520 - Remove unnecessary manual clearing of _initialized_paths in test (implementation already handles deleted databases via exists() check)

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/unit/test_token_service.py [hunk 2]
```diff
@@ -384,31 +397,138 @@ def test_has_scope_hierarchy_is_correct(self, token_service: TokenService) -> No
...
+
+        # Verify the token works
+        assert service.validate_token(token1) is not None
+
+        # Delete the database file
+        test_config.database_path.unlink()
+        assert not test_config.database_path.exists()
+
+        # Force re-initialization by clearing the class-level tracking
+        TokenService._initialized_paths.discard(test_config.database_path)
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/unit/test_token_service.py]: Fixed in 649faab. Removed the manual `_initialized_paths.discard()` call - the test now verifies the actual production behavior where the `exists()` check on line 120 of service.py handles deleted databases automatically. Test passes without any artificial setup, confirming the implementation works correctly.

(AI-generated via Claude Code w/ Claude Sonnet 4.5)

**Comment by duplico** [comment]
## Round 1 Complete

**Comments addressed**: 1/1
**Commits**: 649faab

### Actions Taken
- Removed unnecessary manual clearing of _initialized_paths in test_database_reinitializes_if_deleted
- Test now verifies actual production behavior where exists() check handles deleted databases
- All 49 unit tests in test_token_service.py pass

### Results
- Commit: 649faab4c82b78d96c6370c5ec3a54458068b081
- Reply posted to review comment ID 2750258773

### Remaining
- 0 unresolved threads (after resolving this one)
- Next: Wait for CI, then adversarial review

(AI-generated via Claude Code w/ Sonnet 4.5)

**Comment by duplico** [comment]
CI pending. Waiting for checks to complete before adversarial review.

(AI-generated via Claude Code w/ Sonnet 4.5)

**Comment by duplico** [comment]
CI passed. Starting adversarial review.

(AI-generated via Claude Code w/ Sonnet 4.5)

**Comment by duplico** [comment]
## Adversarial Review: PASS

No significant correctness concerns found. PR is ready for merge.

### Verification Chain
- Issue #164 requirement: Lazy DB initialization to avoid init when TokenService is unused
- PR implementation: Correctly implements lazy init via _ensure_database_initialized()
- Code verification: All DB-accessing methods properly call lazy init
- Tests: Comprehensive coverage including thread safety and edge cases

### Integration Impact
- Backward compatible (API surface unchanged)
- magpie-ctl init still works (explicit init before TokenService usage)
- FastAPI singleton via @lru_cache ensures single instance across requests
- token list CLI correctly bypasses TokenService (intentional)

### Thread Safety
- Double-checked locking pattern correctly implemented
- Class-level _initialized_paths set prevents duplicate initialization
- Handles deleted database files via exists() check
- Test coverage: 10-thread concurrent initialization test

**Status**: merge_ready

(AI-generated via Claude Code w/ Opus 4.5)

(AI-generated via Claude Code w/ Opus 4.5)

**Comment by duplico** [comment]
## Clarity Improvement

Added clear documentation to the double-checked locking logic in `_ensure_database_initialized()` to explain the intentional asymmetry:

**Fast path (line 120):** Conservative - requires BOTH `path in set AND exists()` to skip initialization
- Only skips if we're CERTAIN the DB is ready
- Avoids acquiring the lock in the common case

**Slow path (line 126):** Liberal - reinitializes if `path NOT in set OR NOT exists()`
- Re-initializes if there's ANY doubt
- Handles external deletion or missed initialization without risk

The conditions are intentionally asymmetric. This prevents future maintainers from "simplifying" the logic into a bug by making them symmetric negations of each other.

**Testing:**
- All 849 unit tests pass
- Lint, security, e2e checks pass
- No behavioral changes - pure documentation improvement

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/magpie/auth/service.py [hunk 3]
```diff
@@ -121,6 +151,9 @@ def create_token(self, name: str, scope: TokenScope, plaintext_token: str | None
             TokenFormatError: If provided token has incorrect format.
             ValidationError: If token name fails validation (invalid format/length).
         """
+        # Ensure database is initialized before any operations
+        self._ensure_database_initialized()
+
         # Validate token name (defense in depth - also validated at API layer)
         validate_token_name(name)
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/auth/service.py]: Valid observation. The current pattern initializes the database before validation for consistency — all 4 public methods use the same init-first approach. This makes reasoning about the class simpler: every public method guarantees the DB exists.

The trade-off (invalid requests creating an empty DB file on first call) is low-impact since init_database() is idempotent and the DB will be needed for legitimate operations anyway.

Tracked in a follow-up issue for future consideration.

(AI-generated via Claude Code w/ Haiku 4.5)

**Comment by duplico** [comment]
## Adversarial Review Follow-Up Complete

Addressed findings from deeper adversarial review:

### Issues Analyzed
- **CRITICAL 1 (TOCTOU)**: FALSE POSITIVE - Code was correct. Added documentation explaining GIL-protected operations.
- **CRITICAL 2 (Instance isolation)**: FALSE POSITIVE - Design was correct. Added test and documentation to clarify.
- **HIGH 1 (has_scope())**: FALSE POSITIVE - Method doesn't call lazy init (pure comparison).
- **HIGH 2 (Concurrent test)**: VALID - Enhanced test with barrier synchronization and 20 threads.
- **HIGH 3 (Cache invalidation)**: VALID - Added clear_token_service_cache() for test scenarios.

### Changes Made
1. Enhanced documentation explaining thread safety guarantees
2. Added test_multiple_instances_with_different_paths() to verify isolation
3. Improved concurrent test with threading.Barrier and increased thread count
4. Added cache clearing mechanism for test use

### Results
- All 850 unit tests passing
- All CI checks passing (lint, security, test, e2e)
- Code quality maintained

See `.ai/scratch/adversarial-review-analysis.md` for detailed analysis.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/unit/test_token_service.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (duplico) [reply tests/unit/test_token_service.py]: The lazy initialization machinery (including _initialized_paths, _init_lock, and double-checked locking) was completely removed in commit c1bf54f. The PR now uses a simple @lru_cache singleton pattern for database initialization, so this concern no longer applies.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/magpie/auth/service.py [hunk 4]
```diff
@@ -94,8 +114,41 @@ def __init__(self, config: MagpieSettings) -> None:
...
+        Double-checked locking conditions (intentionally asymmetric):
+        - Fast path: Conservative - only skip if BOTH tracked AND exists on disk.
+          This avoids acquiring the lock when we're certain the DB is ready.
+        - Slow path: Liberal - reinitialize if EITHER not tracked OR file deleted.
+          This handles external deletion or missed initialization without risk.
+
+        Thread safety notes:
+        - The outer check reads from _initialized_paths (a set) without holding the lock.
+          In CPython, set membership checks ('in' operator) and Path.exists() are atomic
+          operations protected by the GIL, making the fast-path read safe.
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/auth/service.py]: The lazy initialization machinery (including _initialized_paths, _init_lock, and double-checked locking) was completely removed in commit c1bf54f. The PR now uses a simple @lru_cache singleton pattern for database initialization, so concerns about GIL atomicity and set operations no longer apply.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/magpie/server/deps.py [hunk 5]
```diff
@@ -31,16 +32,40 @@ def get_storage_service() -> StorageService:
...
-    """Get TokenService instance configured from settings.
+    """Get TokenService instance configured from settings (cached singleton).
+
+    The TokenService is cached to avoid recreating it on every request.
+    This means that changes to MagpieSettings after the first call will
+    not be reflected in the TokenService. In production, settings should
+    be configured once at application startup.
+
+    For testing scenarios where settings change, use clear_token_service_cache()
+    to invalidate the cache before creating a new TokenService with different settings.
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/server/deps.py:45]: Fixed in 184798f56f5b6f39cfea390ded240365778b316d. Updated the docstring to clarify that clear_token_service_cache() only clears the TokenService cache, and that get_settings() has its own separate cache managed by Pydantic Settings.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/integration/conftest.py [hunk 6]
```diff
@@ -423,6 +423,20 @@ def override_auth_dependencies(request):
...
+    """Clear the lru_cache on get_token_service between tests.
+
+    The get_token_service dependency uses @functools.lru_cache(maxsize=1)
+    to create a singleton. Without clearing this cache between tests, a
+    TokenService instance from one test could leak into another test,
+    breaking test isolation.
+    """
+    get_token_service.cache_clear()
+    yield
+    get_token_service.cache_clear()
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/integration/conftest.py]: Fixed in 184798f56f5b6f39cfea390ded240365778b316d. Changed the test fixture to use clear_token_service_cache() instead of calling get_token_service.cache_clear() directly, making it use the public API.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/magpie/server/deps.py [hunk 7]
```diff
@@ -31,16 +32,44 @@ def get_storage_service() -> StorageService:
...
+
+    Note: This only clears the TokenService cache. If you've changed environment
+    variables and need to pick up new settings, you'll also need to clear the
+    get_settings() cache separately (which is managed by Pydantic Settings).
+
+    Example:
+        >>> clear_token_service_cache()
+        >>> service = get_token_service()  # Creates new instance with current settings
+    """
+    get_token_service.cache_clear()
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/server/deps.py:70]: Fixed in 1e00958. Added TestTokenServiceCaching with tests for singleton identity and cache clearing, mirroring the existing get_settings() cache tests.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/magpie/server/deps.py [hunk 8]
```diff
@@ -31,16 +32,44 @@ def get_storage_service() -> StorageService:
...
+    """Clear the cached TokenService instance.
+
+    This is primarily useful in testing scenarios where settings change
+    and a fresh TokenService instance is needed. In production, settings
+    should be configured once at application startup, so this should not
+    be necessary.
+
+    Note: This only clears the TokenService cache. If you've changed environment
+    variables and need to pick up new settings, you'll also need to clear the
+    get_settings() cache separately (which is managed by Pydantic Settings).
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/server/deps.py]: Fixed in a4a64f4. Updated docstring to correctly reference @lru_cache and get_settings.cache_clear().

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/unit/test_deps.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    ) -> None:
+        """get_token_service() should return the same cached instance."""
+        # Set up temporary database path
+        monkeypatch.setenv("MAGPIE_DATABASE_PATH", str(tmp_path / "test.db"))
+        get_settings.cache_clear()
+
+        clear_token_service_cache()
+        service1 = get_token_service()
+        service2 = get_token_service()
+        assert service1 is service2
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/unit/test_deps.py:33]: Fixed in 032afd3. Added autouse fixture to clean up token service cache after each test.

(AI-generated via Claude Code w/ Sonnet 4.5)
